### ISSUE TITLE
Consider GraphQL instead of REST

### ISSUE THREAD

**Issue by jooola** [airsonic/airsonic#1388]
So I have started to work on the really old #553 PR. But on my way there I though we should give a try to Graphql. 

Even if I'm not a big fan of big company like Facebook, Graphql is a really powerful technology.

- It provides really fine grained data queries (way easier than REST)
- Majority of all the data served by airsonic is read only (we should not plan on building a tag editor). I would suggest a REST API if our data would have a lot of Create/Update/Delete endpoint but this is not the case.
- No documentation has to be generated or written, a graphql browser can do this for us.
- Many JS library provide a easy way to bind with the API.

So I'll start working once I have a few positive answers from the people around airsonic.
CC @airsonic/maintainers

**Comment by fxthomas** [comment]
I am neutral on this.

I will definitely not have the time to work on it, and I have zero experience with GraphQL, but if it helps with your use case, why not.

Points you may want to consider:
* This means we will support two competing APIs (Subsonic REST and GraphQL).
* Make it as unobtrusive as possible while in development. You want to minimize overhead in testing your work so that people can actually give you feedback easily :wink: 
* I don't think we should merge this unless we have at least one stable use case (or at least beta, I assume you want to use this in Airsonic UI). Otherwise there is a high risk that we'll end up with something that don't use at all.

**Comment by jooola** [comment]
> This means we will support two competing APIs (Subsonic REST and GraphQL).

I don't think the new RESAPI goal was to enhance the existing Subsonic API. More like one primary API (airsonic) and the second (subsonic) only for the application ecosystem based on it.

This was also meant to make the subsonic API read only.
My idea is to build this primary API using Graphql, which is the same work as building the RESTAPI.

> I don't think we should merge this unless we have at least one stable use case (or at least beta, I assume you want to use this in Airsonic UI?). Otherwise there is a high risk that we'll end up with something that we don't use at all.

Yes the big idea is to build a proper UI. I have to admit I worked a bit on the airsonic-ui with bfkesley but the technology we used is such a mess. So I might plans on rewriting it from scratch. But the problem we faced with this project is that the backend haven't moved a bit while we were asking for more features to be implemented. So I want to do this the other way around.

**Comment by fxthomas** [comment]
> This was also meant to make the subsonic API read only

FWIW, I do use some write features (e.g. starring) of the Subsonic REST API all the time :smile: 

> The backend haven't moved a bit while we were asking for more features to be implemented. So I want to do this the other way around.

Go for it, given that we don't have nearly enough drive-by contributors willing to handle things like this it's probably a good approach!

**Comment by jooola** [comment]
> FWIW, I do use some write features (e.g. starring) of the Subsonic REST API all the time smile

My bad, I simplified my meaning: I meant disabling all admin operation. And optionally provide a subsonic api key to those who want to use it.

**Comment by bfkelsey** [comment]
> the technology we used is such a mess
I think the technology used in the ui was fine. The problem was working with the subsonic api was a pain and nobody wanted to do the major changes needed to fix the backend.

**Comment by jooola** [comment]
> I think the technology used in the ui was fine. The problem was working with the subsonic api was a pain and nobody wanted to do the major changes needed to fix the backend.

IMO both were a problem. Maybe the technology we used wasn't meant for me, so this makes is subjective.

**Comment by muff1nman** [comment]
IMO the closer we can get the api to the actual data, the easier it will be to interact with. The subsonic api seems foreign compared to our actual database. I'm not familiar enough with graphql to say whether it will help or hinder, so maybe you can do a PR show and tell for us :)

> I don't think we should merge this unless we have at least one stable use case (or at least beta, I assume you want to use this in Airsonic UI?). Otherwise there is a high risk that we'll end up with something that we don't use at all.

I agree with @fxthomas and would additionally say the risk of a PR like this becoming stale / not getting merged is high. Maybe just keep that in mind so that you don't put too much work into it before everyone gets a chance to give you early feedback.

**Comment by jooola** [comment]
> I agree with @fxthomas and would additionally say the risk of a PR like this becoming stale / not getting merged is high. Maybe just keep that in mind so that you don't put too much work into it before everyone gets a chance to give you early feedback.

I guess this is the same idea for #553. Will do some a little setup will see what it takes.

**Comment by randomnicode** [comment]
> 
> > I don't think we should merge this unless we have at least one stable use case (or at least beta, I assume you want to use this in Airsonic UI?). Otherwise there is a high risk that we'll end up with something that we don't use at all.
> 
> I agree with @fxthomas and would additionally say the risk of a PR like this becoming stale / not getting merged is high. Maybe just keep that in mind so that you don't put too much work into it before everyone gets a chance to give you early feedback.

I think with big projects like these, a major problem is that the size of the PR starts getting unwieldy. The PR is in no state to get merged, but at the same time the diffs start piling up and harder to scrutinize. And moreover collaboration over a PR is difficult unless you fork the repo and collaboration happens there, or folks keep hijacking a 'running' PR and submitting another PR with the former PR incorporated, which then becomes the current 'running' PR.

My suggestion, if stuff starts getting big enough, is actually to give these fine gentlemen a branch forked off master that they may use to collaborate, submit PRs and merge to. FWIW, it's how I've seen most feature development done with git, especially research projects or things with extended timelines. At some point in the distant future, once the feature is complete and ready to merge into master, we can just merge the feature branch back into master. In the meantime people work off the branch, not cluttering master as well as not having to parse individual changes as part of a massive PR. If the feature becomes stale and ancient, or goes nowhere, no biggie, we just delete the branch. Stale branches, unlike stale PRs don't show up the same way, and don't clutter the active PRs view. Note the PRs will have to be submitted against that branch.

As for GraphQL, much like @fxthomas and @muff1nman I'm fairly ambivalent, though I am curious why you'd exclude 'write' operations?

I have no background regarding this discussion, so I'm also curious who we're actually designing an API for? Is there an additional ecosystem or apps to take advantage of a new API? Or are we hoping a nice new API will spark an ecosystem?

Maybe someone should work on clients too then...
